### ISSUE TITLE
Improve speed of SP local inhibition, 70x slower than global 

### ISSUE THREAD

**Issue by breznak** [htm-community/nupic.cpp#123]
CC CUDA SP #50, SP faster LocalInho #86 , #92  

From https://github.com/htm-community/nupic.cpp/pull/119#pullrequestreview-178811874 : 

I'm evaluation the performance impact of ND vs 1D, where algorithmically the 1D is really simpler. 

- Global Inh 
```
[ RUN      ] HelloSPTPTest.performance
starting test. DIM_INPUT=10000, DIM=2048, CELLS=10
EPOCHS = 1000
starting:  1000 iterations.Total elapsed time = 0 seconds
[       OK ] HelloSPTPTest.performance (1027 ms)
[----------] 1 test from HelloSPTPTest (1027 ms total)
```

- Local, 1D (to be removed): 
```
[ RUN      ] HelloSPTPTest.performance
starting test. DIM_INPUT=10000, DIM=2048, CELLS=10
EPOCHS = 1000
starting:  1000 iterations.Total elapsed time = 74 seconds
/home/mmm/devel/HTM/htm-community/nupic.cpp/src/test/unit/algorithms/HelloSPTPTest.cpp:101: Failure
Value of: timeTotal <= CI_avg_time
  Actual: false
Expected: true
HelloSPTP test slower than expected! (74,should be 45
[  FAILED  ] HelloSPTPTest.performance (74195 ms)
[----------] 1 test from HelloSPTPTest (74195 ms total)
```

- local, ND (proposed default, resp. only choice): 
```
[ RUN      ] HelloSPTPTest.performance
starting test. DIM_INPUT=10000, DIM=2048, CELLS=10
EPOCHS = 1000
starting:  1000 iterations.Total elapsed time = 75 seconds
/home/mmm/devel/HTM/htm-community/nupic.cpp/src/test/unit/algorithms/HelloSPTPTest.cpp:101: Failure
Value of: timeTotal <= CI_avg_time
  Actual: false
Expected: true
HelloSPTP test slower than expected! (75,should be 45
[  FAILED  ] HelloSPTPTest.performance (75335 ms)
[----------] 1 test from HelloSPTPTest (75335 ms total)
```

TL;DR
- local inhibition is very slow! 70x (1s vs 74s global/local inhibition) 
- this PR (1D/ND) does not change that (-> :ok: merge) 
- we should work on improving speed of local inh (#50 ) 
- there are no unit tests for local inhibition!

**Comment by breznak** [comment]
Thanks to PR #769 local inhibition (only for `wrapAround=true`) is now 2x faster! 

> local inhibition is very slow! 70x (1s vs 74s global/local inhibition) 

which equals to improvement and it's now only 10x slower than global inh!
